File: 프로그래머스/unrated/142085. 디펜스 게임/디펜스 게임.kt
import java.util.PriorityQueue;
import java.util.Collections;

class Solution {
    fun solution(n: Int, k: Int, enemy: IntArray): Int {
        var N = n
        var K = k
        var cleared: PriorityQueue<Int> = PriorityQueue<Int>(Collections.reverseOrder())
        var answer: Int = 0
        var cur: Int = 0
        while(N >= 0 && cur < enemy.size){
            N -= enemy[cur]
            cleared.offer(enemy[cur])
            if(N < 0 && K > 0){
                N += cleared.poll()
                K -= 1
            }
            if (N >= 0){
                cur += 1
            }
        }
        return cur
    }
}

File: 프로그래머스/unrated/142085. 디펜스 게임/디펜스 게임.py
import heapq

def solution(n, k, enemy):
    cur = 0
    cleared = []
    answer = 0
    while(n >= 0 and cur < len(enemy)):
        n -= enemy[cur]
        heapq.heappush(cleared, -enemy[cur])
        if (n < 0 and k > 0):
            max_cleared = -1 * heapq.heappop(cleared)
            n += max_cleared
            k -= 1
        if (n >= 0):
            answer += 1
            cur += 1
    return answer